### ISSUE TITLE
Rack reboots during build cause EOF and forever "building" state

### ISSUE THREAD

**Issue by ddollar** [convox/rack#72]
```
$ convox deploy
... a bunch of stuff
scikit-learn-0  81% |#########################      | Time: 0:00:00  39.05 MB/s
scikit-learn-0  86% |##########################     | Time: 0:00:00  39.93 MB/s
scikit-learn-0  91% |############################   | Time: 0:00:00  40.80 MB/s
ERROR: Get https://REDACTED/system: EOF

$ convox releases
ERROR: Get https://REDACTED/system: EOF

$ convox deploy

$ convox logs
ERROR: websocket.Dial wss://REDACTED/apps/app/logs: unexpected EOF

... later

$ convox builds
BGYMFSPRXHU  building               22 minutes ago

```

**Comment by ddollar** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;XNU04rSI&#x2F;177-rack-reboots-break-builds"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Rack reboots break builds</a>

**Comment by nzoschke** [comment]
This needs some design...  A couple ideas:

* Don't allow rack updates during a build
* Don't allow builds during a rack update
* Define a hard timeout, past which builds display as "failed" not "building"

**Comment by mtrofimm** [comment]
I have a similar  issue and in my case it's seems that there is a out-of-memory problem in the docker build process, for example assetic:dump in symofy2 app is very heavy.

**Comment by nzoschke** [comment]
Testing this after the build refactor. You can disrupt the API with:

```
$ convox api get /apps/convox/processes | jq -r '.[] | select(.name == "web") | .id' | xargs -I{} -L1 convox ps stop {} --app convox
Stopping c63d041d14f1... OK
Stopping f41a3d430b0f... OK
```

Doing this during a build results in:

```
$ cd simple-rails

$ convox build
Creating tarball... 
OK
Uploading... 
OK
RUNNING: tar xz
RUNNING: docker build -t convox-mbtcelxstb /go/src/github.com/convox/rack/src
Sending build context to Docker daemon   212 kB
Step 1 : FROM rails:4.2
4.2: Pulling from library/rails
70e9a6907f10: Pulling fs layer
...
2016/04/09 18:47:45 exit status 2
ERROR: response status: 500
```

The build gets terminated, and the record is still stuck in created:

```
$ convox builds
ID           STATUS    RELEASE      STARTED        ELAPSED  DESC
BPOHRKSKMWO  created                4 minutes ago 
```

**Comment by nzoschke** [comment]
I am hoping with a bit more work we can:

* Fully fork the build process from the API process
* Have cmd/build call back to the API with its log data
* Have cmd/build retry the callback a few times

In this fashion, starting a build kicks off a container totally isolated from the Rack API and more reliably updates its final state.

Losing an instance that a build is happening on is not recoverable, but we can better report this as a timeout / failure too.

**Comment by ddollar** [comment]
Fixed in #1231
